src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.neuralsearch.processor;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+
+import lombok.extern.log4j.Log4j2;
+
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.search.TotalHits;
+import org.opensearch.neuralsearch.search.CompoundTopDocs;
+
+/**
+ * Abstracts combination of scores in query search results.
+ */
+@Log4j2
+public class ScoreCombiner {
+
+    private static final Float ZERO_SCORE = 0.0f;
+
+    /**
+     * Performs score combination based on input combination technique. Mutates input object by updating combined scores
+     * @param queryTopDocs query results that need to be normalized, mutated by method execution
+     * @param combinationTechnique exact combination method that should be applied
+     * @return list of max combined scores for each shard
+     */
+    public List<Float> combineScores(final CompoundTopDocs[] queryTopDocs, final ScoreCombinationTechnique combinationTechnique) {
+        List<Float> maxScores = new ArrayList<>();
+        for (int i = 0; i < queryTopDocs.length; i++) {
+            CompoundTopDocs compoundQueryTopDocs = queryTopDocs[i];
+            if (Objects.isNull(compoundQueryTopDocs) || compoundQueryTopDocs.totalHits.value == 0) {
+                maxScores.add(ZERO_SCORE);
+                continue;
+            }
+            List<TopDocs> topDocsPerSubQuery = compoundQueryTopDocs.getCompoundTopDocs();
+            int shardId = compoundQueryTopDocs.scoreDocs[0].shardIndex;
+            Map<Integer, float[]> normalizedScoresPerDoc = new HashMap<>();
+            int maxHits = 0;
+            TotalHits.Relation totalHits = TotalHits.Relation.EQUAL_TO;
+            for (int j = 0; j < topDocsPerSubQuery.size(); j++) {
+                TopDocs topDocs = topDocsPerSubQuery.get(j);
+                int hits = 0;
+                for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                    if (!normalizedScoresPerDoc.containsKey(scoreDoc.doc)) {
```

Comment:
changed, but not sure if result is better, it's mainly because it's not possible to generate array of floats and fill it in one call, and putIfAbsent doesn't take the functional inteface.